st2common/st2common/logging/formatters.py

```diff
@@ -0,0 +1,209 @@
+import time
+import json
+import traceback
+
+import six
+
+__all__ = [
+    'ConsoleLogFormatter',
+    'GelfLogFormatter'
+]
+
+HOSTNAME = socket.gethostname()
+COMMON_ATTRIBUTE_NAMES = [
+    'process',
+    'processName',
+    'module',
+    'filename',
+    'funcName',
+    'lineno'
+]
+
+
+class BaseExtraLogFormatter(logging.Formatter):
+    """
+    Base class for the log formatters which expect additional context to be passed in the "extra"
+    dictionary.
+
+    For example:
+
+    extra={'_id': 'user-1', '_path': '/foo/bar'}
+
+    Note: To avoid clashes with standard Python log record attributes, all the keys in the extra
+    dictionary need to be prefixed with a slash ('_').
+    """
+
+    PREFIX = '_'
+
+    def _get_extra_attributes(self, record):
+        attributes = dict([(k, v) for k, v in six.iteritems(record.__dict__)
+                           if k.startswith(self.PREFIX)])
+        return attributes
+
+    def _get_common_extra_attributes(self, record):
+        result = {}
+
+        for name in COMMON_ATTRIBUTE_NAMES:
+            value = getattr(record, name, None)
+
+            if not value:
+                continue
+
+            result[name] = value
+
+        return result
+
+    def _format_extra_attributes(self, attributes):
+        simple_types = (list, dict, int, float) + six.string_types
+
+        result = {}
+        for key, value in six.iteritems(attributes):
+            if isinstance(value, simple_types):
+                # Leave simple types as is
+                value = value
+            elif isinstance(value, object):
+                if getattr(value, 'to_dict', None):
+                    # Check for a custom serialization method
+                    value = value.to_dict()
+                else:
+                    value = repr(value)
+
+            result[key] = value
+
+        return result
+
+
+class ConsoleLogFormatter(BaseExtraLogFormatter):
+    """
+    Formatter which attaches all the attributes from the "extra" dictionary as key=value pairs to
+    the end of the log message.
+
+    For example:
+
+        LOG.info('Hello world', extra={'_id': 1, '_path': '/fooo'})
+
+    Result:
+
+        Hello World (id=1,path=foo)
+    """
+
+    def format(self, record):
+        attributes = self._get_extra_attributes(record=record)
+        attributes = self._format_extra_attributes(attributes=attributes)
+        attributes = self._dict_to_str(attributes=attributes)
+
+        # Call the parent format method so the final message is formed based on the "format"
+        # attribute in the config
+        msg = super(ConsoleLogFormatter, self).format(record)
+
+        if attributes:
+            msg = '%s (%s)' % (msg, attributes)
+
+        return msg
+
+    def _dict_to_str(self, attributes):
+        result = []
+        for key, value in six.iteritems(attributes):
+            item = '%s=%s' % (key[1:], repr(value))
+            result.append(item)
+
+        result = ','.join(result)
```

Comment:
umm. not sure how that whitespace kills grep but i'll leave this to you. it's minor.